rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport.h

```diff
@@ -35,25 +252,555 @@ namespace rmw_fastrtps_cpp
+                            subros_message = align_(max_align, 0, subros_message, space);
+                        }
+                    }
+                    break;
+                default:
+                    printf("unknown type id %u\n", member->type_id_);
+                    throw std::runtime_error("unknown type");
+            }
+        }
+    }
+
+    return true;
+}
+
+template <typename MembersType>
+bool rmw_fastrtps_cpp::TypeSupport<MembersType>::deserializeROSmessage(
+    eprosima::fastcdr::Cdr &deser, const MembersType *members, void *ros_message, bool call_new)
+{
+    assert(members);
+    assert(ros_message);
+
+    for(unsigned long i = 0; i < members->member_count_; ++i)
+    {
+        const auto *member = members->members_ + i;
+        void *field = (char*)ros_message + member->offset_;
+
+        if(!member->is_array_)
+        {
+            switch(member->type_id_)
+            {
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_BOOL:
+                    {
+                        deser >> *(bool*)field;
+                    }
+                    break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_BYTE:
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_UINT8:
+                    deser >> *(uint8_t*)field;
+                    break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_CHAR:
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_INT8:
+                    deser >> *(char*)field;
+                    break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_FLOAT32:
+                    deser >> *(float*)field;
+                    break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_FLOAT64:
+                    deser >> *(double*)field;
+                    break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_INT16:
+                    deser >> *(int16_t*)field;
+                    break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_UINT16:
+                    deser >> *(uint16_t*)field;
+                    break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_INT32:
+                    deser >> *(int32_t*)field;
+                    break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_UINT32:
+                    deser >> *(uint32_t*)field;
+                    break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_INT64:
+                    deser >> *(int64_t*)field;
+                    break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_UINT64:
+                    deser >> *(uint64_t*)field;
+                    break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_STRING:
+                    {
+                        std::string &str = *(std::string*)field;
```

Comment:
Yeah, that sounds more like it :-) I still don't know why the examples work, maybe I was running the listener for another implementation by mistake. It'd be great if you can work on this, I added a templated struct that returns the appropriate type depending on whether we're using the C or the C++ structures, it might be useful in this case:

https://github.com/eProsima/ROS-RMW-Fast-RTPS-cpp/pull/18/files#diff-a8e9397d37ed8274594931575b5c83e1R211